src/hash/blake/mod.rs

```diff
@@ -0,0 +1,261 @@
+use super::{ElementHasher, HashFn};
+use crate::{
+    ByteReader, ByteWriter, Deserializable, DeserializationError, Digest, Felt, FieldElement,
+    Serializable,
+};
+use core::{
+    mem::{size_of, transmute},
+    ops::Deref,
+    slice::from_raw_parts,
+};
+
+#[cfg(test)]
+mod tests;
+
+// BLAKE3 256-BIT OUTPUT
+// ================================================================================================
+
+/// 256-bit output of a blake3 function.
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq)]
+pub struct Blake3Digest256([u8; Self::SIZE]);
+
+impl Blake3Digest256 {
+    /// Bytes length of the output.
+    pub const SIZE: usize = 32;
+}
+
+impl Deref for Blake3Digest256 {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Serializable for Blake3Digest256 {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_u8_slice(&self.0);
+    }
+}
+
+impl Deserializable for Blake3Digest256 {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        source.read_u8_array().map(Self)
+    }
+}
+
+impl Digest for Blake3Digest256 {
+    fn as_bytes(&self) -> [u8; 32] {
+        self.0
+    }
+}
+
+/// 256-bit output blake3 hasher.
+pub struct Blake3_256;
+
+impl HashFn for Blake3_256 {
+    type Digest = Blake3Digest256;
+
+    fn hash(bytes: &[u8]) -> Self::Digest {
+        Blake3Digest256(blake3::hash(bytes).into())
+    }
+
+    fn merge(values: &[Self::Digest; 2]) -> Self::Digest {
+        Self::hash(prepare_merge(values))
+    }
+
+    fn merge_with_int(seed: Self::Digest, value: u64) -> Self::Digest {
+        let mut hasher = blake3::Hasher::new();
+        hasher.update(&seed.0);
+        hasher.update(&value.to_le_bytes());
+        Blake3Digest256(hasher.finalize().into())
+    }
+}
+
+impl ElementHasher for Blake3_256 {
+    type BaseField = Felt;
+
+    fn hash_elements<E>(elements: &[E]) -> Self::Digest
+    where
+        E: FieldElement<BaseField = Self::BaseField>,
+    {
+        Self::hash(E::elements_as_bytes(elements))
```

Comment:
Unfortunately, I don't think we can hash elements as is. The reason is that this will hash the internal representation of elements (which, in our case, is Montgomery form). So, this could produce un-intuitive results when we has a sequence of bytes but then the same sequence of bytes represented by a field element would hash to something else.

Reply:
That is correct; my bad there. Fixed!